pkg/controller/daemon/daemon_controller.go

```diff
@@ -865,12 +884,29 @@ func (dsc *DaemonSetsController) podsShouldBeOnNode(
 				continue
 			}
 			if pod.Status.Phase == v1.PodFailed {
+				failedPodsObserved++
+
+				// This is a critical place where DS is often fighting with kubelet that rejects pods.
+				// We need to avoid hot looping and backoff.
+				backoffKey := failedPodsBackoffKey(ds, node.Name)
+
+				now := dsc.failedPodsBackoff.Clock.Now()
+				inBackoff := dsc.failedPodsBackoff.IsInBackOffSinceUpdate(backoffKey, now)
+				if inBackoff {
+					delay := dsc.failedPodsBackoff.Get(backoffKey)
+					glog.V(4).Infof("Deleting failed pod %s/%s on node %s has been limited by backoff - %v remaining",
+						pod.Namespace, pod.Name, node.Name, delay)
+					dsc.enqueueDaemonSetAfter(ds, delay)
+					continue
+				}
+
+				dsc.failedPodsBackoff.Next(backoffKey, now)
```

Comment:
Do we need `failedPodsBackoff`? Why not just use `dsc.enqueueDaemonSetRateLimited(ds)`?

Reply:
Yes, we need it. Re-queuing rate-limited does not prevent other events on the DaemonSet causing it to be processed prior to that rate limited re-add, which would trigger delete/recreate of all of the failed pods on the failed nodes.  We want the controller to remain responsive, without churning on permanently failing nodes